sub-packages/bionemo-dnadl/src/bionemo/dnadl/io/dataset.py

```diff
@@ -0,0 +1,189 @@
+from bionemo.dnadl.tools.genome_interval import GenomeInterval
+from bionemo.dnadl.tools.vcf import SampleId, read_variants_in_interval
+from bionemo.noodles.nvfaidx import NvFaidx
+
+
+DNATokenizer = Callable[[str], torch.Tensor]
+
+
+class Genome:
+    """Class that creates a genome object from a fasta file. It can be used to extract sequences from the genome."""
+
+    def __init__(self, fasta_file: str | os.PathLike):
+        """Instantiate the class."""
+        fasta_file = Path(fasta_file)
+        assert fasta_file.exists(), "path to fasta file must exist"
+        self.seqs = NvFaidx(str(fasta_file))
+
+    def adjust_interval(self, interval: GenomeInterval, context_length: int) -> GenomeInterval:
+        """Used to extend an interval by a fixed amount on both sides."""
+        start, end = interval.start, interval.end
+        interval_length = interval.end - interval.start
+        chromosome_length = self.seqs[interval.chromosome].length
+        if interval_length >= context_length:
+            return interval
+
+        extra_seq = context_length - interval_length
+        extra_left_seq = extra_seq // 2
+        extra_right_seq = extra_seq - extra_left_seq
+
+        start -= extra_left_seq
+        end += extra_right_seq
+
+        if start <= 0:
+            start = 1
+
+        if end > chromosome_length:
+            end = chromosome_length
+
+        return GenomeInterval(interval.chromosome, start, end)
+
+    def extract_sequence(self, genome_interval: GenomeInterval) -> str:
+        """Extract a sequence from the genome."""
+        return str(self.seqs[genome_interval.chromosome][genome_interval.start : genome_interval.end])
+
+
+class GenomeDataset(Dataset, ABC):
+    """Abstract class for datasets that return a DNA sequence."""
+
+    def __init__(self, tokenizer: DNATokenizer):
+        """Instantiate the class."""
+        self.tokenizer = tokenizer
+
+    def __getitem__(self, ind: int) -> torch.tensor:
+        """Get an item from the dataset."""
+        sequence = self.get_dna_sequence(ind)
+        return self.tokenizer(sequence)
+
+    def get_dna_sequence(self, ind: int) -> str:
+        """Get the DNA sequence."""
+        ...
+
+
+class GenomeIntervalDataset(GenomeDataset):
+    """Datasets that extract sequences from a genome based on a bed file."""
+
+    def __init__(
+        self,
+        genome: Genome,
+        bed_file_df: pl.DataFrame,
+        context_length: int | None = None,
+        chr_bed_to_fasta_map: dict = {},
+        **kwargs: Any,
+    ):
+        """Instantiate the class.
+
+        Args:
+            genome: Genome object
+            bed_file_df: A bed file in the form of a polars DataFrame with chrom, start, and end columns at least
+            context_length: Size of the window
+            chr_bed_to_fasta_map: Mapping between chromosome names in the bed file and the fasta file
+            **kwargs: Additional arguments
+        """
```

Comment:
I'm confused here. Context-length is used to set the window size, but the sequence returned seems to be a function of both the interval size- which could vary within the bedfile, the context length, (and potentially the end of the contig). Is this intentional? Or maybe more specifically, are there restrictions on the contents of `bed_file_df` that youre expecting?

If I were to use this dataset out of the box, depending on the model, I think I would need to transform the outputs to have a fixed size, with padding or truncation.

Reply:
This is actually from the previous BioNeMo, but i think I know where it comes from because we also have this issue in-house. Back in the day when language modelling wasn't a thing in DNA there weren't tokenizers and tokens (e.g. Enformer). you one hot encoded the sequence. for a batch you needed to have windows of the same size, and you didn't have PAD tokens. It was more biologically relevant to include more of the actual DNA sequence than add Ns (if the context_length was longer than the intervals specified in your file), so you had to deal with this at the Dataset level, rather than at the point of collation.. The name is not vey informative, I can change to force_window_size but let's align on a strategy here.